### DISCUSSION TITLE
sharing-appsettings-json-configuration-files-between-projects-in-asp-net-core/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#99] [Announcements]
# sharing-appsettings-json-configuration-files-between-projects-in-asp-net-core/

In this post I show how you can extract common settings to a shared JSON file and how to configure your projects to use them locally and after publishing

https://andrewlock.net/sharing-appsettings-json-configuration-files-between-projects-in-asp-net-core/

**Comment by gsingh203** [comment]
Thanks for the article.  Very helpful.  Do I need to create two SharedSettings file if I have settings different for Dev and Prod environments.

**Comment by RichardHowells** [comment]
I'm trying this idea out.  I like it so far.

I found one issue - for Console projects then looking upwards only one level of directory from the executable does not find the SharedSettings.json file.  I'm trying out having the SharedSettings.json link in the project set to BOTH CopyToPublishDirectory AND CopyToOutput.  That way in both debug AND running from a publish it's in the same directory as the executable.

  <ItemGroup>
    <Content Include="..\SharedSettings.json" Link="SharedSettings.json" CopyToPublishDirectory="Always">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

The CopyToPublishDirectory has to be manually edited into the project file.  The CopyToOutputDirectory can be set using the VS GUI.

**Comment by dooie** [comment]
Add file as link might be a simpler approach.

Add your file in the root then in the projects you want the file 'right click the project file > add > existing item > add as link'

more info [Here](https://medium.com/entech-solutions/share-config-files-between-web-console-and-other-types-of-net-core-projects-87fa40b4e65e)

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 06 2018, 09:24, in reply to [Koste](https://disqus.com/by/disqus_Yo3mbtFAOi/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/99#discussioncomment-6045735)</em>

---

<p>That's certainly possible, but I wanted to avoid the case where I have multiple duplicate instances of the files in the projects. I realise your approach would update them every build, but I've used a similar approach in the past and always found myself editing the wrong file, and getting my changes wiped out! Definitely makes the AddJsonFile scripts easier though, and is a good solution if it works for you :)</p><p>I've fixed the link so you can give it a go, thanks.</p>

[a comment by an automated account is not shown]

**Reply by gsingh203** [reply]
I am using the same approach you mentioned.  In the shared project I set each sharedsettings.{environment}.json to copy always.  I build the shared project.  But the files don't get copied into the bin folder of the project that references the shared project.  The files only copy to the bin folder of the shared project.  Not sure what I am missing.

[a comment by an automated account is not shown]

**Comment by NickSt** [comment]
Late post for this, but I had to solve a similar problem recently and I that I think the best way to handle this is to assign one app as the "owner" and have it publish the appsettings.json to the shared location if doesn't exist or is out of date via static method before loading the configuration in your program cs file.  This way there are no conflicts as long as both apps are not trying to do something like write to the file.  The only downside to this, is that it requires that the appsettings owner has run at least once before the second application can run.  You can handle this with a while loop that checks for the file before allowing app2 to start though.

**Comment by lnicolae** [comment]
?? In MyApp.Api use the file; in MyApp.Web use a shortcut to the file. One file shared between projects.
